File: src/utils/calculators/sds-calculator.ts
import type { Answer } from '../../types'

export interface SDSResult extends Record<string, any> {
  rawScore: number
  standardScore: number
  percentile: number
  level: 'normal' | 'mild' | 'moderate' | 'severe'
  levelText: string
  dimensions: {
    affective: { score: number; percentage: number; level: string }
    somatic: { score: number; percentage: number; level: string }
    cognitive: { score: number; percentage: number; level: string }
    psychomotor: { score: number; percentage: number; level: string }
  }
  radarData: { dimension: string; score: number; fullMark: number }[]
  interpretation: string
  recommendations: string[]
  warning: string
}

export function calculateSDS(answers: Answer[]): SDSResult {
  const dimensionMap: Record<string, string[]> = {
    affective: ['sds_n01', 'sds_n02', 'sds_n03', 'sds_n15'],
    somatic: ['sds_n04', 'sds_n05', 'sds_n07', 'sds_n08', 'sds_n09', 'sds_n10'],
    cognitive: ['sds_n11', 'sds_n14', 'sds_n16', 'sds_n17', 'sds_n18', 'sds_n19'],
    psychomotor: ['sds_n06', 'sds_n12', 'sds_n13', 'sds_n20'],
  }

  const reverseScoredItems = ['sds_n02', 'sds_n05', 'sds_n06', 'sds_n11', 'sds_n12', 'sds_n14', 'sds_n16', 'sds_n17', 'sds_n18', 'sds_n20']
  
  const answerMap: Record<string, number> = {}
  answers.forEach(a => {
    let value = typeof a.value === 'number' ? a.value : parseInt(String(a.value || 1))
    if (reverseScoredItems.includes(a.questionId)) {
      value = 5 - value
    }
    answerMap[a.questionId] = value
  })

  const dimensions: SDSResult['dimensions'] = {} as SDSResult['dimensions']
  let totalRaw = 0

  Object.entries(dimensionMap).forEach(([dim, ids]) => {
    const score = ids.reduce((sum, id) => sum + (answerMap[id] || 2), 0)
    const maxScore = (ids.length || 1) * 4
    const percentage = Math.round((score / maxScore) * 100)
    let level = '正常'
    if (percentage >= 75) level = '重度'
    else if (percentage >= 60) level = '中度'
    else if (percentage >= 50) level = '轻度'
    dimensions[dim as keyof SDSResult['dimensions']] = { score, percentage, level }
    totalRaw += score
  })

  const standardScore = Math.round(totalRaw * 1.25)
  let percentile = 50
  if (standardScore < 53) percentile = 25
  else if (standardScore < 63) percentile = 50
  else if (standardScore < 73) percentile = 75
  else percentile = 95

  let level: SDSResult['level'] = 'normal'
  let levelText = '正常范围'
  if (standardScore >= 73) { level = 'severe'; levelText = '重度抑郁' }
  else if (standardScore >= 63) { level = 'moderate'; levelText = '中度抑郁' }
  else if (standardScore >= 53) { level = 'mild'; levelText = '轻度抑郁' }

  const radarData = [
    { dimension: '情感障碍', score: dimensions.affective.percentage, fullMark: 100 },
    { dimension: '躯体症状', score: dimensions.somatic.percentage, fullMark: 100 },
    { dimension: '认知偏差', score: dimensions.cognitive.percentage, fullMark: 100 },
    { dimension: '精神运动', score: dimensions.psychomotor.percentage, fullMark: 100 },
  ]

  const interpretations = {
    normal: '你的情绪状态整体健康，没有明显抑郁症状。继续保持积极的生活态度和健康的生活方式。',
    mild: '你可能处于轻度抑郁状态。建议多与朋友家人交流，适当增加户外活动，注意情绪调节。',
    moderate: '你有中度抑郁倾向。建议寻求专业心理咨询，学习情绪管理技巧，建立健康的作息规律。',
    severe: '你有较明显的抑郁症状。强烈建议立即寻求专业心理医生或精神科医师的帮助，请不要独自面对。',
  }

  const recommendations = [
    '每天保持 30 分钟以上的户外活动，晒太阳有助于血清素分泌',
    '建立规律的作息时间，保证 7-8 小时睡眠',
    '适度运动，有氧运动是天然的抗抑郁剂',
    '减少社交媒体使用时间，避免信息过载',
    '每周至少与朋友深度交流一次',
  ]

  const warnings = level === 'severe' 
    ? '⚠️ 本测评仅作初步筛查，不能替代专业诊断。如果有自杀念头，请立即拨打心理援助热线。'
    : level === 'moderate'
    ? '⚠️ 如果症状持续超过两周，建议寻求专业心理咨询帮助。'
    : ''

  return {
    rawScore: totalRaw,
    standardScore,
    percentile,
    level,
    levelText,
    dimensions,
    radarData,
    interpretation: interpretations[level],
    recommendations,
    warning: warnings,
  }
}

File: backend/app/services/calculators/sds_calculator.py
from typing import Dict, Any, List
from .base import BaseCalculator


class SDSCalculator(BaseCalculator):
    def calculate(self, answers: List[Dict[str, Any]]) -> Dict[str, Any]:
        total_score = sum(answer.get("value", 0) for answer in answers)
        question_count = len(answers)
        
        if question_count == 0:
            return {"total_score": 0, "level": "normal"}

        mean_score = total_score / question_count
        
        if mean_score >= 4:
            level = "severe"
        elif mean_score >= 3:
            level = "moderate"
        elif mean_score >= 2:
            level = "mild"
        else:
            level = "normal"

        return {
            "total_score": total_score,
            "mean_score": round(mean_score, 2),
            "level": level,
            "question_count": question_count,
        }

    def get_interpretation(self, result: Dict[str, Any]) -> Dict[str, Any]:
        level = result.get("level", "normal")
        level_descriptions = {
            "normal": {
                "title": "正常范围",
                "description": "您的抑郁水平在正常范围内，情绪状态良好。",
                "suggestion": "继续保持积极的生活态度和健康习惯。",
            },
            "mild": {
                "title": "轻度抑郁",
                "description": "您可能偶尔感到情绪低落或失去兴趣，这是正常的情绪波动。",
                "suggestion": "关注自己的情绪变化，与亲友多交流。",
            },
            "moderate": {
                "title": "中度抑郁",
                "description": "您的抑郁情绪较为明显，可能影响到日常生活和工作。",
                "suggestion": "建议寻求专业心理咨询帮助。",
            },
            "severe": {
                "title": "重度抑郁",
                "description": "您的抑郁水平较高，可能严重影响到日常生活和身心健康。",
                "suggestion": "强烈建议立即寻求专业心理健康服务。",
            },
        }
        return level_descriptions.get(level, level_descriptions["normal"])

    def get_recommendations(self, result: Dict[str, Any]) -> List[str]:
        level = result.get("level", "normal")
        
        if level == "severe":
            return [
                "立即寻求专业心理健康服务",
                "告知亲友您的状况",
                "避免独处，保持社交联系",
                "进行适度的户外活动",
            ]
        elif level == "moderate":
            return [
                "寻求心理咨询师帮助",
                "建立规律的日常生活",
                "保持适度运动",
                "与信任的人分享感受",
            ]
        elif level == "mild":
            return [
                "保持规律作息",
                "进行喜欢的活动",
                "与朋友保持联系",
                "练习正念或冥想",
            ]
        else:
            return [
                "继续保持积极的生活态度",
                "保持社交活动",
                "定期自我检查情绪状态",
            ]
